Add wireguard server

runesoerensen opened the pull request:
Sets up some of the more boring parts of the system that future pull requests (the API and CLI) rely on.

Note that the nginx process has been moved to a separate container, and the wireguard-server container now just routes traffic to it. Let me know if you have thoughts on that or anything else here :)

runesoerensen committed:
Add docker compose and with nginx service

runesoerensen committed:
Add wireguard-server docker container

runesoerensen committed:
Add wireguard-server service to docker compose

runesoerensen committed:
Configure routing from wireguard network to nginx container

runesoerensen committed:
Save peer configuration when container is stopped

klizhentas commented on wireguard-server/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+EXPOSE 51820/udp
+
+# Set a simple static authentication token shared with the Wirescale API
+ENV WGREST_STATIC_AUTH_TOKEN=secret
```
This means it will be visible in the env of the docker, which is anti-pattern, better use secret:

https://docs.docker.com/engine/swarm/secrets/

Or use file mount

runesoerensen replied:
For sure -- I was planning to use a file mount or similar when there was a need for sharing it between containers (i.e. when I add the code that uses it in the Wirescale API).

Definitely not pretty to have this in the env and I've pushed a change to use secrets instead. Thanks!

runesoerensen committed:
Use Docker secrets to manage shared auth token for wgrest

The wgrest_auth_token.txt file containing the secret wouldn't typically be committed to version control, but added it here for convenience

marcoandredinis commented on wireguard-server/entrypoint.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
```
💅 
```suggestion
#!/usr/bin/env bash
```

marcoandredinis commented on wireguard-server/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM debian:bullseye-slim
+
+RUN apt-get update && apt-get install -y wireguard iptables iproute2 curl systemd
```
That's usually a good practice to also `apt upgrade` before installing anything
Also, is `systemd` required? I assume we might need it as a dependency for something else but not directly required.

runesoerensen replied:
Added the apt upgrade before installing!

runesoerensen replied:
About `systemd`: It'd definitely be cleaner to not install it -- I did so initially because the wgrest package uses it in a post-installation script (e.g. https://github.com/suquant/wgrest/blob/master/packaging/nfpm-amd64.yaml). Probably easy to avoid this. Any suggestions? :)

marcoandredinis replied:
No problem 👍 You can leave it as is
It was mostly me being curious about why we need systemd when we are in a container environment.
It will be hard to have `systemd` as PID 1 and I don't think it will run otherwise.
But it looks like we need it because of it creating a specific folder, so that's ok to install it 👍

runesoerensen replied:
Ok will leave it as is for now then, but it sure is awkward to run systemd here :)

Thanks for all your help and feedback here, much appreciated!

marcoandredinis commented on wireguard-server/wg-wirescale.conf:
```diff
@@ -0,0 +1,5 @@
+PrivateKey = <server_private_key>
+Address = 10.8.0.1/24
+ListenPort = 51820
+SaveConfig = true
```
Setting `SaveConfig` to true and doing `wg-quick save wg-wirescale` looks to be equivalent 🤔 
If so, should we remove one?

runesoerensen replied:
You're right -- the `wg-quick save wg-wirescale` hack is currently used specifically because this configuration doesn't have any effect. Would be nice with a cleaner solution, but I've removed `SaveConfig = true` for now :)

marcoandredinis reviewed (approved):
Just left some comments
No blockers, so approving

marcoandredinis commented on wireguard-server/install-wgrest.sh:
```diff
@@ -0,0 +1,21 @@
+    exit 1
+    ;;
+  esac
+  curl -L https://github.com/suquant/wgrest/releases/latest/download/wgrest_$ARCH.deb -o wgrest_amd.deb
+  dpkg -i wgrest_amd.deb
```
We should pin the version when downloading the dependency

We should also ask curl to throw an error when there's an HTTP error (eg 500, 404).
This ensures we get an error instead of writing gibberish to the file.
```
  wgrest_version=1.0.0-alpha10.5
  # ...
  curl -fssL https://github.com/suquant/wgrest/releases/download/${wgrest_version}/wgrest_$ARCH.deb -O
  dpkg -i wgrest_$ARCH.deb
```

runesoerensen replied:
Fixed - thanks!

runesoerensen committed:
Merge branch 'main' into add-wireguard-server

runesoerensen committed:
Make entrypoint.sh shebang more portable

Co-authored-by: Marco André Dinis <marcoandredinis@gmail.com>

runesoerensen committed:
Pin wgrest version and fail fast on http error

runesoerensen committed:
Prefer apt over apt-get and upgrade before installing

runesoerensen committed:
Remove SaveConfig from wireguard interface configuration file
